examples/js/loaders/GLTFLoader.js

```diff
@@ -366,45 +391,82 @@ THREE.GLTFLoader = ( function () {
 	 */
 	function GLTFMaterialsUnlitExtension() {
 
-		this.name = EXTENSIONS.KHR_MATERIALS_UNLIT;
+		this.extension = EXTENSIONS.KHR_MATERIALS_UNLIT;
 
 	}
 
-	GLTFMaterialsUnlitExtension.prototype.getMaterialType = function () {
+	GLTFMaterialsUnlitExtension.prototype = {
 
-		return THREE.MeshBasicMaterial;
+		constructor: GLTFMaterialsUnlitExtension,
 
-	};
+		onMaterial: function ( materialDef, parser ) {
 
-	GLTFMaterialsUnlitExtension.prototype.extendParams = function ( materialParams, materialDef, parser ) {
+			var extensions = parser.plugins.extensions;
+			var materialParams = {};
+			var pending = [];
 
-		var pending = [];
+			materialParams.color = new THREE.Color( 1.0, 1.0, 1.0 );
+			materialParams.opacity = 1.0;
 
-		materialParams.color = new THREE.Color( 1.0, 1.0, 1.0 );
-		materialParams.opacity = 1.0;
+			var metallicRoughness = materialDef.pbrMetallicRoughness;
 
-		var metallicRoughness = materialDef.pbrMetallicRoughness;
+			if ( metallicRoughness ) {
 
-		if ( metallicRoughness ) {
+				if ( Array.isArray( metallicRoughness.baseColorFactor ) ) {
 
-			if ( Array.isArray( metallicRoughness.baseColorFactor ) ) {
+					var array = metallicRoughness.baseColorFactor;
 
-				var array = metallicRoughness.baseColorFactor;
+					materialParams.color.fromArray( array );
+					materialParams.opacity = array[ 3 ];
 
-				materialParams.color.fromArray( array );
-				materialParams.opacity = array[ 3 ];
+				}
 
-			}
+				if ( metallicRoughness.baseColorTexture !== undefined ) {
 
-			if ( metallicRoughness.baseColorTexture !== undefined ) {
+					pending.push( parser.assignTexture( materialParams, 'map', metallicRoughness.baseColorTexture ) );
 
-				pending.push( parser.assignTexture( materialParams, 'map', metallicRoughness.baseColorTexture ) );
+				}
 
 			}
 
-		}
+			return Promise.all( pending ).then( function () {
 
-		return Promise.all( pending );
+				var material = new THREE.MeshBasicMaterial( materialParams );
+
+				if ( materialDef.name !== undefined ) material.name = materialDef.name;
+				if ( materialDef.doubleSided === true ) material.side = THREE.DoubleSide;
+
+				var alphaMode = materialDef.alphaMode || ALPHA_MODES.OPAQUE;
+
+				if ( alphaMode === ALPHA_MODES.BLEND ) {
+
+					material.transparent = true;
+
+				} else {
+
+					material.transparent = false;
+
+					if ( alphaMode === ALPHA_MODES.MASK ) {
+
+						material.alphaTest = materialDef.alphaCutoff !== undefined ? materialDef.alphaCutoff : 0.5;
+
+					}
+
+				}
+
+				// baseColorTexture, emissiveTexture, and specularGlossinessTexture use sRGB encoding.
+				if ( material.map ) material.map.encoding = THREE.sRGBEncoding;
+				if ( material.emissiveMap ) material.emissiveMap.encoding = THREE.sRGBEncoding;
+
+				assignExtrasToUserData( material, materialDef );
+
+				if ( materialDef.extensions ) addUnknownExtensionsToUserData( extensions, material, materialDef );
+
+				return material;
```

Comment:
Refer to [here](https://github.com/mrdoob/three.js/pull/18421#discussion_r370892841) for duplicated code with the one in `.loadMaterial()` which sets up some common properties.

Reply:
I guess such simple addition extensions can be handled with `.onAfterXXX`. (In those material simple addition cases also with `.onBeforeCompile` or node-based material if three.js core doesn't support those properties.)

The duplicated code problem I described above can be happened if (a.) extension handler makes an object different type from core spec object and (b.) the handler want to avoid the cost of creation/download core spec dependent assets. So I chose material unlit extension as the first extension moving to the new system to discuss.

Reply:
I'm leaning toward 2. option. And also moving `assignExtrasToUserData()` and `addUnknownExtensionsToUserData()` out from `.loadXXX()`, and calling in `getDependency()` instead because these two functions may be necessary to be called regardless of handlers.

We can rewrite to

```javascript
//.getDependency()
case 'material':
  var materialDef = json.materials[index];
  dependency = this._onBefore(type, materialDef).then(function (def) {
    materialDef = def;
    return parser._on(type, materialDef);
  }).then(function (material) {
    return material || parser.loadMaterial(materialDef);
  }).then(function (material) {
    material = parser._onAfter(type, material, materialDef);
    assignExtrasToUserData(material, materialDef);
    if (materialDef.extensions) addUnknownExtensionsToUserData(extensions, material, materialDef);
    return material;
  });
  break;
 
// better method name?
GLTFParser.prototype.finalizeMaterial = function (material, materialDef) {
  if (materialDef.name !== undefined) material.name = materialDef.name;
  if (materialDef.doubleSided === true) material.side = THREE.DoubleSide;

  var alphaMode = materialDef.alphaMode || ALPHA_MODES.OPAQUE;
  if (alphaMode === ALPHA_MODES.BLEND) {
    material.transparent = true;
  } else {
    material.transparent = false;
    if (alphaMode === ALPHA_MODES.MASK) {
      material.alphaTest = materialDef.alphaCutoff !== undefined ? materialDef.alphaCutoff : 0.5;
    }
  }

  // baseColorTexture, emissiveTexture, and specularGlossinessTexture use sRGB encoding.
  if (material.map) material.map.encoding = THREE.sRGBEncoding;
  if (material.emissiveMap) material.emissiveMap.encoding = THREE.sRGBEncoding;
  return material;
};

GLTFParser.prototype.loadMaterial = function (materialDef) {
  var material = new THREE.MeshStandardMaterial();
  ....
  return this.finalizeMaterial(material, materialDef);
};

GLTFMaterialsUnlitExtension.prototype.onMaterial = function (materialDef, parser) {
  var material = new THREE.MeshBasicMaterial();
  ....
  return this.finalizeMaterial(material, materialDef);
};
```

Reply:
Moved `assignExtrasToUserData()` and `addUnknownExtensionsToUserData()` calls from `.loadXXX()` into `.getDependency()` so far.

Reply:
Coming to wonder if exposing `.finalizeXXX()` is the best idea because the parser can be complex by having bunch of relatively small methods... I'd like to think of the solution through more extension handlers. I added an inline note that material unlit extension handler currently has common property setup code duplicated with the one in `.loadMaterial()` so far and want to optimize later if needed. What do you think?

Reply:
Making `assignTexture` non material specific probably sounds like a good idea. One concern is taking encoding and format parameters enables to refer the same texture but with different parameters, like

```javascript
pending.push(parser.assignTexture(materialParams, 'fooMap', mapDef, THREE.RGBFormat));
pending.push(parser.assignTexture(materialParams, 'barMap', mapDef, THREE.RGBAFormat));
```

in such a case parser needs to clone texture. So we need to add code maintaining it in `assignTexture`.

I want to think of this adding encoding and format parameters in another PR (if needed) because I don't really want to make this PR bigger.

Reply:
Let's not change `assignTexture` right now, agreed.